Source/Classes/PINRemoteImageDownloadQueue.m

```diff
@@ -99,42 +100,58 @@ - (void)scheduleDownloadsIfNeeded
                 [task resume];
                 
-                _runningOperationCount++;
+                [_runningTasks addObject:task];
             }
         }
     [self unlock];
 }
 
-- (void)removeDownloadTaskFromQueue:(NSURLSessionDataTask *)downloadTask
+- (BOOL)removeDownloadTaskFromQueue:(NSURLSessionDataTask *)downloadTask
 {
+    BOOL containsTask = NO;
     [self lock];
-        [_highPriorityQueuedOperations removeObject:downloadTask];
-        [_defaultPriorityQueuedOperations removeObject:downloadTask];
-        [_lowPriorityQueuedOperations removeObject:downloadTask];
+        if ([_highPriorityQueuedOperations containsObject:downloadTask]) {
+            containsTask = YES;
+            [_highPriorityQueuedOperations removeObject:downloadTask];
+        } else if ([_defaultPriorityQueuedOperations containsObject:downloadTask]) {
+            containsTask = YES;
+            [_defaultPriorityQueuedOperations removeObject:downloadTask];
+        } else if ([_lowPriorityQueuedOperations containsObject:downloadTask]) {
+            containsTask = YES;
+            [_lowPriorityQueuedOperations removeObject:downloadTask];
+        }
     [self unlock];
+    return containsTask;
 }
 
 - (void)setQueuePriority:(PINRemoteImageManagerPriority)priority forTask:(NSURLSessionDataTask *)downloadTask
 {
-    [self removeDownloadTaskFromQueue:downloadTask];
+    [self setQueuePriority:priority forTask:downloadTask addIfNecessary:NO];
+}
+
+- (void)setQueuePriority:(PINRemoteImageManagerPriority)priority forTask:(NSURLSessionDataTask *)downloadTask addIfNecessary:(BOOL)addIfNecessary
+{
+    BOOL containsTask = [self removeDownloadTaskFromQueue:downloadTask];
     
-    NSMutableArray <NSURLSessionDataTask *> *queue = nil;
-    [self lock];
-        switch (priority) {
-            case PINRemoteImageManagerPriorityLow:
-                queue = _lowPriorityQueuedOperations;
-                break;
-                
-            case PINRemoteImageManagerPriorityDefault:
-                queue = _defaultPriorityQueuedOperations;
-                break;
-                
-            case PINRemoteImageManagerPriorityHigh:
-                queue = _highPriorityQueuedOperations;
-                break;
-        }
-        [queue addObject:downloadTask];
-    [self unlock];
+    if (containsTask || addIfNecessary) {
+        NSMutableArray <NSURLSessionDataTask *> *queue = nil;
+        [self lock];
+            switch (priority) {
+                case PINRemoteImageManagerPriorityLow:
+                    queue = _lowPriorityQueuedOperations;
+                    break;
+                    
+                case PINRemoteImageManagerPriorityDefault:
+                    queue = _defaultPriorityQueuedOperations;
+                    break;
+                    
+                case PINRemoteImageManagerPriorityHigh:
+                    queue = _highPriorityQueuedOperations;
+                    break;
+            }
```

Comment:
What is happening on the default case? Should we assert in case an invalid priority is passed in.